File: 1101-1199/1145/1145.c
#include <stdio.h>

int n, m, i, j, k, l, size;

void bfs(int i, int t[])
{
	int q1[size], q2[size];
	q1[0] = i;
	q2[0] = 0;

	int s=0, e=1;
	t[i] = 0;
	int x = 0;
	while (s < e)
	{
		j = q1[s];
		l = q2[s];
		s++;
		if (j%n != 0 && t[j-1] == 1)
		{
			t[j-1] = 0;
			q1[e] = j-1;
			q2[e] = l+1;
			e++;
		}
		if (j%n != n-1 && t[j+1] == 1)
		{
			t[j+1] = 0;
			q1[e] = j+1;
			q2[e] = l+1;
			e++;
		}
		if (j >= n && t[j-n] == 1)
		{
			t[j-n] = 0;
			q1[e] = j-n;
			q2[e] = l+1;
			e++;
		}
		if (j <= size-n-1 && t[j+n] == 1)
		{
			t[j+n] = 0;
			q1[e] = j+n;
			q2[e] = l+1;
			e++;
		}
	}
}

int main()
{
	scanf("%d %d", &n, &m);
	
	size = n*m;
	int a[size], b[size];
	char ch;
	
	for (i=0; i<m; i++)
	{
		for (j=0; j<n; j++)
		{
			scanf("%c", &ch);
			while (ch != '.' && ch != '#')
				scanf("%c", &ch);
			if (ch == '.')
			{
				k = i*n+j;
				a[k] = 1;
				b[k] = 1;
			}
			else
			{
				a[i*n+j] = 0;
				b[i*n+j] = 0;
			}	
		}
	}
	
	bfs(k, a);
	int f1 = l;
	bfs(j, b);
	int f2 = l;
	printf("%d", (f1 > f2) ? f1 : f2);
	return 0;
}

File: 1101-1199/1145/1145.py
import sys

n, m = [int(x) for x in sys.stdin.readline().split()]
a = [False]*(n*m)
b = [False]*(n*m)
k1 = -1

for i in range(m):
	line = sys.stdin.readline()
	for j in range(n):
		if line[j] == '.':
			k1 = i*n+j
			a[k1] = True
			b[k1] = True

def bfs(i, t):
	q1 = [-1]*(n*m)
	q2 = [-1]*(n*m)
	q1[0], q2[0] = i, 1
	s, e = 0, 1
	t[i] = False
	while s < e:
		j, l = q1[s], q2[s]
		s += 1
		if j%n != 0 and t[j-1]:
			q1[e] = j-1
			q2[e] = l+1
			e += 1
			t[j-1] = False
		if j%n != n-1 and t[j+1]:
			q1[e] = j+1
			q2[e] = l+1
			e += 1
			t[j+1] = False
		if j >= n and t[j-n]:
			q1[e] = j-n
			q2[e] = l+1
			e += 1
			t[j-n] = False
		if j <= (m-1)*n-1 and t[j+n]:
			q1[e] = j+n
			q2[e] = l+1
			e += 1
			t[j+n] = False
	return j, l

k2, d1 = bfs(k1, a)
k3, d2 = bfs(k2, b)

print max(d1, d2)-1
	
